staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,570 @@
+	// We prefer the simplicity over the promptness, at least for now.
+
+	// Create a request and enqueue
+	req := &fq.Request{
+		DequeueChannel:  make(chan bool, 1),
+		RealEnqueueTime: qs.clock.Now(),
+		Queue:           queue,
+	}
+	if ok := qs.rejectOrEnqueue(req); !ok {
+		return nil
+	}
+	metrics.ObserveQueueLength(qs.config.Name, len(queue.Requests))
+	return req
+}
+
+// removeTimedOutRequestsFromQueue rejects old requests that have been enqueued
+// past the requestWaitLimit
+func (qs *queueSet) removeTimedOutRequestsFromQueue(queue *fq.Queue) {
+	timeoutIdx := -1
+	now := qs.clock.Now()
+	reqs := queue.Requests
+	// reqs are sorted oldest -> newest
+	// can short circuit loop (break) if oldest packets are not timing out
+	// as newer packets also will not have timed out
+
+	// now - requestWaitLimit = waitLimit
+	waitLimit := now.Add(-qs.config.RequestWaitLimit)
+	for i, req := range reqs {
+		if waitLimit.After(req.RealEnqueueTime) {
+			qs.counter.Add(1)
+			req.DequeueChannel <- false
+			close(req.DequeueChannel)
+			// get index for timed out packets
+			timeoutIdx = i
+		} else {
+			break
+		}
+	}
+	// remove timed out packets from queue
+	if timeoutIdx != -1 {
+		// timeoutIdx + 1 to remove the last timeout req
+		removeIdx := timeoutIdx + 1
+		// remove all the timeout packets
+		queue.Requests = reqs[removeIdx:]
+		// decrement the # of requestsEnqueued
+		qs.numRequestsEnqueued -= removeIdx
+	}
+}
+
+// getRequestsExecutingLocked gets the # of requests which are "executing":
+// this is the# of requests/packets which have been dequeued but have not had
+// finished (via the FinishRequest method invoked after service)
+func (qs *queueSet) getRequestsExecutingLocked() int {
+	total := 0
+	for _, queue := range qs.queues {
+		total += queue.RequestsExecuting
+	}
+	return total
+}
+
+// chooseQueueIndexLocked uses shuffle sharding to select a queue index
+// using the given hashValue and the shuffle sharding parameters of the queueSet.
+func (qs *queueSet) chooseQueueIndexLocked(hashValue uint64) int {
+	bestQueueIdx := -1
+	bestQueueLen := int(math.MaxInt32)
+	// DesiredNum is used here instead of numQueues to omit quiescing queues
+	qs.dealer.Deal(hashValue, func(queueIdx int) {
+		thisLen := len(qs.queues[queueIdx].Requests)
+		if thisLen < bestQueueLen {
+			bestQueueIdx, bestQueueLen = queueIdx, thisLen
+		}
+	})
+	return bestQueueIdx
+}
+
+// updateQueueVirtualStartTime updates the virtual start time for a queue
+// this is done when a new packet is enqueued.  For more info see:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#dispatching
+func (qs *queueSet) updateQueueVirtualStartTime(packet *fq.Request, queue *fq.Queue) {
+	// When a request arrives to an empty queue with no requests executing:
+	// len(queue.Requests) == 1 as enqueue has just happened prior (vs  == 0)
```

Comment:
I don't understand this comment.

Reply:
The hope of this comment was to help explain why the check was at  `len(queue.Requests) == 1` and not `len(queue.Requests) == 0`.  The KEP reads:

`When a request arrives to an empty queue with no requests executing, the queue’s virtual start time is set to now().`

But the logic has this happening at `len(queue.Requests) == 1` (not 'an empty queue')

Reply:
Please make the comment explain *why* this is to be done? ("it's in the KEP" is not an explanation!)